### PR TITLE
add typescript implementation of Do.try pattern

### LINKED ISSUE #47: Port over Typescript implementation of Do.Try pattern
A Typescript implementation of the `Do.Try` pattern was recently added to one of the client projects. Could be pretty useful to port over into this project.

### PR DESCRIPTION
**Pull request by mrjones2014**
- [x] Related GitHub issue(s) linked in PR description
- [x] Destination branch merged, built and tested with your changes
- [x] Code formatted and follows best practices and patterns
- [x] Code builds cleanly (no _additional_ warnings or errors)
- [x] Manually tested
- [x] Automated tests are passing
- [x] No _decreases_ in automated test coverage
- [x] Documentation updated (readme, docs, comments, etc.)
- [x] Localization: No hard-coded error messages in code files (_minimally_ in string constants)

Fixes #47

### CODE REVIEW

File: src/tests/factories/result-error-record-factory.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import { ResultErrorRecord } from "../..//view-models/result-error-record";
```
Comment (brandongregoryscott) [inline src/tests/factories/result-error-record-factory.ts:1]: This path contains a double slash - looks like it's not failing anything, but it doesn't seem right. Could you quick update this, @mrjones2014?

File: src/utilities/do-try.test.ts [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+PolyfillUtils.registerPromiseFinallyPolyfill();
+
+describe("do-try.ts", () => {
+    describe("Do.try", () => {
+        it("When validation errors occur (i.e. error is a ResultRecord), then passes typed errors to catch handler", async () => {
+            // Arrange
+            const workload = async () => {
+                throw ResultRecordFactory.build({
+                    errors: [ResultErrorRecordFactory.build()],
```
Comment (brandongregoryscott) [inline src/utilities/do-try.test.ts:19]: I believe there's a `buildList` function from rosie, isn't there?

File: src/utilities/do-try.test.ts [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+            // Arrange
+            const workload = () => {
+                throw Error();
+            };
+
+            // Act
+            const result = DoSync.try(workload)
+                .catch((result?: ResultRecord<any>, error?: any) => {
+                    expect(result).toBeUndefined();
+                    expect(error).not.toBeUndefined();
```
Comment (brandongregoryscott) [inline src/utilities/do-try.test.ts:126]: Minor, but there is a `toBeNil()` matcher which could reduce the assertions here. That should handle `undefined` or `null`. Not a big deal, I don't think we've been using it, but I was pretty sure there was a matcher that did both 😛

File: src/utilities/do-try.ts [hunk 4]
```diff
@@ -0,0 +1,179 @@
+import { ResultRecord } from "../view-models/result-record";
+
+// -----------------------------------------------------------------------------------------
+// #region Utility Types
+// -----------------------------------------------------------------------------------------
+
+type AsyncWorkload<T> = () => Promise<T>;
+type SyncWorkload<T> = () => T;
+type CatchHandler<T> = (result?: ResultRecord<T>, error?: any) => void;
```
Comment (brandongregoryscott) [inline src/utilities/do-try.ts:9]: Could these types be broken out into the `types` folder?

**Review by brandongregoryscott — CHANGES_REQUESTED** [review]
Just some minor cleanup @mrjones2014 👍

File: src/tests/factories/result-error-record-factory.ts (hunk 1 above)
Comment (mrjones2014) [reply src/tests/factories/result-error-record-factory.ts:1]: oops 😬 I'll fix it

File: src/utilities/do-try.test.ts (hunk 2 above)
Comment (mrjones2014) [reply src/utilities/do-try.test.ts:19]: neat!

File: src/utilities/do-try.test.ts (hunk 3 above)
Comment (mrjones2014) [reply src/utilities/do-try.test.ts:126]: hm, I'm not seeing that matcher anywhere.

Comment (mrjones2014) [reply src/utilities/do-try.test.ts:126]: ah, looks like it does exist, just VS Code intellisense doesn't pick it up.

Comment (brandongregoryscott) [reply src/utilities/do-try.test.ts]: Yeah, we still need to iron out the typings for `jest-extended`

**Comment by brandongregoryscott** [comment]
@mrjones2014 One last note: you'll want to export these classes/types from `index.ts` or they won't be included in the build for a consumer. I can take care of that in the build/publishing process, just a heads up for next time.

**Review by brandongregoryscott — APPROVED** [review]

**Pull request merged**